Feature/redeem holo fuel form

mateuszRybczonek opened the pull request:
This PR:

- enables "Redeem HoloFuel" button when redeemable HF is above 0 HF on Dashboard -> HoloFuel card
<img width="398" alt="Screenshot 2023-04-25 at 14 25 54" src="https://user-images.githubusercontent.com/17565389/234275497-bee2f059-bd4f-4076-840b-11a074b240a1.png">

- enables "Redeem HoloFuel" button when redeemable HF is above 0 HF on Earnings Page
<img width="1201" alt="Screenshot 2023-04-25 at 14 26 57" src="https://user-images.githubusercontent.com/17565389/234275761-d3ba4287-f1c1-43f4-bd62-ba82aa1ab6cd.png">

- redirects to Redeem HoloFuel form when any of those buttons gets clicked

- adds Redeem HoloFuel Form with input validations
<img width="1467" alt="Screenshot 2023-04-25 at 14 27 41" src="https://user-images.githubusercontent.com/17565389/234275916-b9840882-6e50-45c5-acba-226a3b6d3f86.png">
 
<img width="1489" alt="Screenshot 2023-04-25 at 14 28 28" src="https://user-images.githubusercontent.com/17565389/234276091-dfc7f102-def7-4e56-bec2-67df6754a994.png">
<img width="1501" alt="Screenshot 2023-04-25 at 14 28 54" src="https://user-images.githubusercontent.com/17565389/234276205-480597cf-2f4d-4644-817a-2000a296f1d7.png">

mateuszRybczonek committed:
feat(redeem-form): implement route and basic card layout

mateuszRybczonek committed:
feat(redeem-form): adjust earnings page redeem HoloFuel button

mateuszRybczonek committed:
feat(redeem-form): implement validation on step one and basic step two template

mateuszRybczonek committed:
feat(redeem-form): step two can submit logic and mobile tolerant styles

mateuszRybczonek committed:
feat(redeem-form): disable buttons when redeemable HF is 0

mateuszRybczonek committed:
feat(redeem-form): update submodule

JettTech commented on src/components/dashboard/HoloFuelCard.vue:
```diff
@@ -68,29 +39,70 @@ const items = computed((): DashboardCardItem[] => [
         ? formatCurrency(Number(props.data.redeemable))
         : 0,
-    isActive: false
+    isActive: true
```
```suggestion
    isActive: !!props.data.redeemable
```

Shouldn't the `isActive` be dependant on whether the there is a redeemable amount?

mateuszRybczonek replied:
No, this is for the label and value, not for the button. Those should be always enabled now, but... I will remove the whole property as this was needed only as a temporary solution until the feature is implemented. Thanks

JettTech commented on src/components/earnings/RedeemHoloFuelFormStepOne.vue:
```diff
@@ -0,0 +1,141 @@
+)
+
+watch(
+  () => canSubmit.value,
+  () => {
+    if (canSubmit.value) {
+      emit('update', { amount: amount.value, hotAddress: hotAddress.value })
+    }
+  }
```
Why not combine these into one watch function?

mateuszRybczonek replied:
It is a single watcher. That's how we write them in composition API

```
watch(
  () => propertyToWatch, // source
  (value) => { // callback
    // logic
  }
)
```

JettTech commented on src/components/earnings/RedeemHoloFuelCard.vue:
```diff
@@ -0,0 +1,177 @@
+const router = useRouter()
+
+const props = defineProps<{
+  redeemableHoloFuel: string | number
```
Is there a reason we need to ever have these be a number?  We're not doing any calculations ...and this value will also need to able to take up to 18 decimal digit values, so it would be prob be bet practice to keep it as a string to avoid any awkward js rounding / math - similar to how we're managing the reserve amounts in springboard.

mateuszRybczonek replied:
It can be a string as a prop (updated), but we need to convert it to a number anyway to format it as a currency and to check if the provided value is lower than redeemable value.

<img width="651" alt="Screenshot 2023-04-28 at 13 27 38" src="https://user-images.githubusercontent.com/17565389/235135867-81da02fa-2018-4d28-b047-696b3c78463d.png">

JettTech replied:
Ah, ok, it would be necessary for the value comparison.  

Also... this is a bit of a moot point, but a thought I'd like to confirm nonetheless. -- As far as the currency formatting, I thought as long we set the input type to number and pattern/lang to parse it as a local string we would still get correctly formatted currency strings.  Can you confirm this as true?

JettTech reviewed (approved):
Left a few comments that need to be resolved before merging  ..but otherwise it all lgtm. :)

JettTech reviewed (commented):
Also... I'd like to follow-up on tests.  Are you planning on adding the tests for this feature in the following PR?

mateuszRybczonek committed:
feat(redeem-form): update ui-common-library

mateuszRybczonek committed:
feat(redeem-form): apply review comments

- remove isActive prop from HoloFuel card items
- type redeemableHoloFuel prop as string
- update validation

mateuszRybczonek commented:
We have very limited tests in host-console-ui, that's on my list, but I need to find a proper approach first for testing. It is kind of hard to implement any e2e tests due to very specific API (holoport) that I am not sure how we could mock. I could use the mocked data, but then the test would only run locally against that server. I can definitely add more unit tests for the components though.
